File: backend/src/types/statement.ts
export interface Statement {
  accountNumber: string;
  description: string;
  endBalance: number;
  mutation: number;
  reference: number;
  startBalance: number;
}

export const isStatement = (obj: unknown): obj is Statement => {
  if (!obj) return false;
  if (typeof obj !== "object") return false;
  if (!("accountNumber" in obj) || typeof obj.accountNumber !== "string") return false;
  if (!("description" in obj) || typeof obj.description !== "string") return false;
  if (!("endBalance" in obj) || typeof obj.endBalance !== "number") return false;
  if (!("mutation" in obj) || typeof obj.mutation !== "number") return false;
  if (!("reference" in obj) || typeof obj.reference !== "number") return false;
  if (!("startBalance" in obj) || typeof obj.startBalance !== "number") return false;
  return true;
};

File: statement-processor/src/main/kotlin/rabo/statementprocessor/Statement.kt
package rabo.statementprocessor

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement

@JsonPropertyOrder("reference", "accountNumber", "description", "startBalance", "mutation", "endBalance" )
class Statement(
    @JsonProperty("accountNumber") val accountNumber: String,
    @JsonProperty("description") val description: String,
    @JsonProperty("endBalance") val endBalance: Float,
    @JsonProperty("startBalance") val startBalance: Float,
    @JsonProperty("mutation") val mutation: Float,
    @JsonProperty("reference") val reference: Long
)

@JacksonXmlRootElement(localName = "Transfer")
class WrappedStatement(
    @JacksonXmlProperty(localName="record") val record: List<Statement>
)
